test/threads.jl

```diff
@@ -141,3 +141,41 @@ end
 @test_throws TypeError Atomic{Bool}
 @test_throws TypeError Atomic{BigInt}
 @test_throws TypeError Atomic{Float64}
+
+# Test atomic fences
+type CommBuf
+    barrier::Atomic{Int}
+    data::Vector{Float64}
+    correct::Bool
+    CommBuf() = new(Atomic{Int}(0), Vector{Float64}(100), false)
+end
+function test_fence_write(commbuf::CommBuf)
+    sleep(0.1)
+    for i in eachindex(commbuf.data)
+        commbuf.data[i] = i
+    end
+    atomic_fence()
+    commbuf.barrier[] = 1
```

Comment:
I don't think that's right.

Observing the result of an atomic release store with an atomic acquire load grantees that any memory operations on the thread performing the release store that happens before the store in source order can be observed on the thread performing the acquire load after the load in source order so in this case the write to the data will be available to the read thread as long as the release store to the barrier is after the write to the data (in source order) and as long as the read of data is after the acquire load. See the last code example in the acquire/release order in this GCC wiki https://gcc.gnu.org/wiki/Atomic/GCCMM/AtomicSync .